### ISSUE TITLE
Autologin / Rememberme for FrontendUser only works on second page load

### ISSUE THREAD

**Issue by kartoffelheinz** [contao/contao#2576]
**Affected version(s)**

Contao 4.9

**Description**
When using the autologin function (Cookie REMEMBERME) in the frontend login module, the actual autologin never works on first, but only on second page load.

**How to reproduce**

1. Create a simple page
2. Create MemberGroup Testgroup
3. Create Member Testmember, allow login and assign Testgroup
4. Insert frontend user login module in pages article
5. Insert text element "Loggedin" and protect it for MemberGroup Testgroup
6. Browse to page with a clean browser session, login as Testmember and tick "remember me / autologin"
7. Text "Loggedin" will show.
8. Let session expire or manually delete PHPSESSION Cookie using your browsers development tools
9. Reload page.

**Expected Result**
Text "Loggedin" will show, login form will be hidden and button Logout shows.

**Actual Result**
Text Loggedin does **not** not show, but login form is hidden and button Logout shows. 
When refreshing the page now, the Text Loggedin will show.

In addition to this, if you also you make the page itself procteded for MemberGroup Testgroup this exception will be logged / shown in dev mode:

`ErrorException:
Warning: implode(): Invalid arguments passed

  at vendor/contao/core-bundle/src/Resources/contao/controllers/FrontendIndex.php:281
  at Contao\FrontendIndex->renderPage(object(PageModel))
     (vendor/symfony/http-kernel/HttpKernel.php:158)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/http-kernel/HttpKernel.php:80)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/http-kernel/Kernel.php:201)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (web/index.php:31)
  at require('.../web/index.php')
     (web/app.php:4)      `

**Comment by aschempp** [comment]
Did you enable any caches for that page?

**Comment by kartoffelheinz** [comment]
No, caching was completely disabled (root page and page). This happens in prod and dev environments. I also deleted var/cache/(dev|prod) to make sure there is no cached precompiled code or such things.

I can reproduce this for any standard installation in contao 4.9, and they log / show (for dev mode) the exception I quoted above.

**Comment by fritzmg** [comment]
> **How to reproduce**
> 
>  1. Create a simple page 
>  2. Create MemberGroup Testgroup 
>  3. Create Member Testmember, allow login and assign Testgroup 
>  4. Insert frontend user login module in pages article 
>  5. Insert text element "Loggedin" and protect it for MemberGroup Testgroup 
>  6. Browse to page with a clean browser session, login as Testmember and tick "remember me / autologin" 
>  7. Text "Loggedin" will show. 
>  8. Let session expire or manually delete PHPSESSION Cookie using your browsers development tools 
>  9. Reload page.

I can confirm the problem in `4.9.x-dev` with the aforementioned reproduction.

> In addition to this, this exception will be logged / shown in dev mode:
> 
> `ErrorException:
> Warning: implode(): Invalid arguments passed

I was unable to confirm this problem.

**Comment by aschempp** [comment]
I remember there being a reason for this that we discussed, but I can't remember what it was. This should especially be true for subrequests, but the example shouldn't have any. The only viable option would be to reload the page on a (rememberme) login, which means there might be an unnecessary http request.

**Comment by fritzmg** [comment]
> I remember there being a reason for this that we discussed, but I can't remember what it was.

Hm, but that would be odd. Why would the login module have the correct information about the currently logged in user, but a content element before or after it would not?

**Comment by aschempp** [comment]
Sorry I need to be more clear, this only applies to ESI requests. The first main request reads the rememberme cookie, starts the session and adds a login session cookie to the main response. ESI subrequests of that response neither have a remember me cookie nor a session cookie (because it was not sent in the initial request).

**Comment by fritzmg** [comment]
Yes, but this particular issue is not about ESI requests anyway. It's just a regular `text` content element.

**Comment by kartoffelheinz** [comment]
@fritzmg Did you use APP_ENV=dev in .env and error_reporting=E_ALL in PHP? Otherwise the warning will be supressed I suppose.

It is unfortunate that debugging became so much harder to do after Contao migrated to Symfony, especially after 4.9. Virtually impossible for a standard web developer that is not highly familiar with Symfony to trace code paths. Most times it is just unclear if functionality is provided by Contao or Symfony :(

But I tried to shed some light on this. It seems that the problem is not that the FrontendUser will not be logged in (that is why the login module will not show) but that Contao "forgets" to assign the users groups. And that is where the warning comes from (Contao assumes for a logged in user that the Group property has been set (and if there are no groups, that is is just an empty array).

**Comment by fritzmg** [comment]
> @fritzmg Did you use APP_ENV=dev in .env and error_reporting=E_ALL in PHP? Otherwise the warning will be supressed I suppose.

Warnings are not suppressed in the `dev` environment (or at least they should not be).

> But I tried to shed some light on this. It seems that the problem is not that the FrontendUser will not be logged in (that is why the login module will not show) but that Contao "forgets" to assign the users groups. And that is where the warning comes from (Contao assumes for a logged in user that the Group property has been set (and if there are no groups, that is is just an empty array).

Sounds plausible :)

**Comment by fritzmg** [comment]
@kartoffelheinz your warning is not related to the content element, but to a protected page, which is not included in your reproduction.

In case of the content element, the root cause is that the `FE_USER_LOGGED_IN` constant is set to `false` if a user is logged in via a `REMEMBERME` cookie. The user object itself does have the correct groups - but the `Controller::isVisibleElement` returns early for `protected` elements, if `FE_USER_LOGGED_IN` is false.

**Comment by aschempp** [comment]
> In case of the content element, the root cause is that the FE_USER_LOGGED_IN constant is set to false if a user is logged in via a REMEMBERME cookie.

That's very interesting! When the Contao Framework is bootet, the constants are set. This happens **before** Symfony Security is initialized. The framework reads the session content and extracts whether a user is present. Thats actually an ugly hack and kinda problematic, but I can't think of another solution (except again, the reload) 🤔

**Comment by fritzmg** [comment]
Why is the framework initialized before Symfony Security anyway?

**Comment by fritzmg** [comment]
Btw. even if `$this->tokenChecker->hasFrontendUser()` would return `true` here:

https://github.com/contao/contao/blob/918884ce919f753e8e7fd68c76396bf92cb5703d/core-bundle/src/Framework/ContaoFramework.php#L194-L203

`FE_USER_LOGGED_IN` would still be set to `false`, because `$this->request->hasPreviousSession()` cannot be true in such a case. So Contao actually currently _requires_ you to reload the page, if you are automatically logged in.

**Comment by kartoffelheinz** [comment]
@fritzmg Thank you for clarifying, yes, I forgot that I also put the page into protected mode for further tests. I added the information to my initial posting.

But is this not part (at the very least a symptom) of the bug, too? Please give it a try and setup a protected page, then using the autologin feature. It will also not work on first page load and in addition should show the aforementioned warning (which led me to believe, that the page does know I'm logged in but has no valid Group property set).

@leofeyer Why do you not think this is a bug? This can't possibly be an intentional feature.

I understand that rearranging the code so that the contao framework will be initialized after Symfony Secutiry (which sounds like the only way to go) might be time consuming. So if  there is no easy other way to fix this bug, I'd like to propose an automatic page reload (provided by the contao core after successful autologin) workaround (only for get requests of course).

**Comment by fritzmg** [comment]
@leofeyer this is definitely a bug, but we need to discuss how to best handle it I suppose. Ideally we can figure out a way to make the constant be correct on the first page load as well.

**Comment by fritzmg** [comment]
> @fritzmg Thank you for clarifying, yes, I forgot that I also put the page into protected mode for further tests. I added the information to my initial posting.
> 
> But is this not part (at the very least a symptom) of the bug, too? Please give it a try and setup a protected page, then using the autologin feature. It will also not work on first page load and in addition should show the aforementioned warning (which led me to believe, that the page does know I'm logged in but has no valid Group property set).

I am still unable to reproduce it. What I _can_ confirm is essentially the same as with the content element: the page will not show up in the navigation, because `Module::renderNavigation` also uses the `FE_USER_LOGGED_IN` constant. But it will be accessible and there is no warning.

**Comment by fritzmg** [comment]
@leofeyer btw. the same issue can be reproduced on https://hostingwerk.de/anmelden for example 😬. If you are logged in via `REMEMBERME`, the navigation menu will still say "Anmelden" instead of "Control Panel":

![Screen Shot 2020-12-17 at 10 34 57](https://user-images.githubusercontent.com/4970961/102476905-94edae80-4053-11eb-895a-12acd1e0d3e8.png)

After refresh:

![Screen Shot 2020-12-17 at 10 35 15](https://user-images.githubusercontent.com/4970961/102476924-9ae38f80-4053-11eb-8447-d396be7dc381.png)

**Comment by leofeyer** [comment]
As discussed in Mumble on December 17th, we want to replace `FE_USER_LOGGED_IN` with `$this->tokenChecker->hasFrontendUser()`. We should then deprecate the constant (add it to the `DEPRECATED.md` file).

**Comment by kartoffelheinz** [comment]
Thank you all for working on this. I understand you have decided on a possible solution. Has there been any progress in implementing this so far? I'd be happy to do a few tests if you already have a PR or otherwise "supposed to work"-fix as a patch or just code snippets which I can copy to our Contao source.

**Comment by fritzmg** [comment]
I'll work on this.

**Comment by kartoffelheinz** [comment]
Thank you for providing with patches. 

I just tried them in latest 4.9, but they do not solve the issue for me. I can still reproduce both parts I mentioned (Login only on second page load / Warning in implode for Line 281 in FrontendIndex.php because of the missing (array) typecast on $this->User->Groups which is not an array at this point).

What am I missing here? Is there anything else I need to do in order to make your fixes work?

**Comment by fritzmg** [comment]
The fix has not been released yet.

**Comment by kartoffelheinz** [comment]
I understand. The closed status was confusion at this stage. Maybe it might be better to keep the issue open then. An issue closing (and master-merged) pull request is widely understood as a verified and working solution.

**Comment by fritzmg** [comment]
That is standard practise (not just in this project). Tickets are closed once work on it is done or an alternative ticket or pull request has been opened. Not when the commit containing the fix is contained within a release tag.

**Comment by kartoffelheinz** [comment]
Allright. Although I do not know closing an issue before a fix has been merged standard practice,since this is not my project I accept whatever workflow you have. 

Will you update this issue again as soon as there is a fix / a solution that actually fixes the issue?

**Comment by bytehead** [comment]
As soon as `4.9.11` is available, the issue should be fixed.

**Comment by fritzmg** [comment]
> Allright. Although I do not know closing an issue before a fix has been merged standard practice,since this is not my project I accept whatever workflow you have.

The fix has already been merged. See the linked PR.

**Comment by kartoffelheinz** [comment]
> The fix has already been merged. See the linked PR.

Okay, maybe we're talking about different things then. I specifically applied all patches from the PR and they did not fix the issue. That was reason for my first comment today. Then you told me, a fix as not been released, so I assumed the PR alone is not enough to fix the issue. Now you tell me the PR is the fix. I'm confused.

I'll wait for 4.9.11 (do you know when it will be released?) and test again.

**Comment by fritzmg** [comment]
Ah, so you did in fact test [`4.9.x-dev`](https://packagist.org/packages/contao/core-bundle#4.9.x-dev)? It should be fixed there. If it is not, there may still be another issue in your case.

**Comment by kartoffelheinz** [comment]
No, I did not test the full branch, since I did not know what exactly I would find in there. Suffix -dev sometimes means different things and is oftentimes filled with errors / bugs. You might use it differently as some kind of pre-master-staging, I suppose. 

At the point I checked and tested, relevant to me was only the fix to the bug I reported, so I tested by applying every change from the PR that fixed this issue (feeding patch the output from github's .patch extension). I gather from your responses that this might not have been enough because there might be needed changes merged into 4.9.x-dev that were not referenced in the commit.

Anyway, I'l just wait for 4.9.11 now. Do you know by any chance when it will be released? Because we suffer from this bug in production :/

**Comment by fritzmg** [comment]
> No, I did not test the full branch, since I did not know what exactly I would find in there. Suffix -dev sometimes means different things and is oftentimes filled with errors / bugs. You might use it differently as some kind of pre-master-staging, I suppose.

Yes, installing the current dev branch should only be done for testing, not for a production environment.

> At the point I checked and tested, relevant to me was only the fix to the bug I reported, so I tested by applying every change from the PR that fixed this issue (feeding patch the output from github's .patch extension). I gather from your responses that this might not have been enough because there might be needed changes merged into 4.9.x-dev that were not referenced in the commit.

No, these changes should be enough. Please test the `4.9x-dev` branch in a testing environment of your site.

**Comment by kartoffelheinz** [comment]
Sorry for the late response. I now found the time to test with 4.9x-dev AND the just released 4.9.11. Unfortunately, neither fixed the issue in our installations.

My initial case for the issue still stands. The first load on a page which triggers the remember me method will NOT show the protected content element (while also NOT showing "guest only" elements), whereas consequent page loads work just fine. And the implode warning for protected pages still occurs (since the array typecast on User->groups is still missing).

What else can I do to help? I do not feel comfortable enough with Symfony/Contao Core to debug the code tough.

**Comment by fritzmg** [comment]
I am unable to reproduce the problem with the reproduction steps of your original post. Can you test it in a fresh Contao 4.9.11 installation and provide new reproduction steps?

**Comment by kartoffelheinz** [comment]
Another testcase: Use the Password Change Module on an unprotected page. Protected the module itself, but not the content element that inserts the module. Login with remember me, browse to said page, delete session cookie, reload. The following error will show (if dev mode is on, otherwise in logfile):

Error:
Call to a member function getTable() on null

  at vendor/contao/core-bundle/src/Resources/contao/modules/ModuleChangePassword.php:110
  at Contao\ModuleChangePassword->compile()
     (vendor/contao/core-bundle/src/Resources/contao/modules/Module.php:214)
  at Contao\Module->generate()
     (vendor/contao/core-bundle/src/Resources/contao/modules/ModuleChangePassword.php:56)
  at Contao\ModuleChangePassword->generate()
     (vendor/contao/core-bundle/src/Resources/contao/elements/ContentModule.php:75)
  at Contao\ContentModule->generate()
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:575)
  at Contao\Controller::getContentElement(object(ContentModel), 'main')
     (vendor/contao/core-bundle/src/Resources/contao/modules/ModuleArticle.php:218)
  at Contao\ModuleArticle->compile()
     (vendor/contao/core-bundle/src/Resources/contao/modules/Module.php:214)
  at Contao\Module->generate()
     (vendor/contao/core-bundle/src/Resources/contao/modules/ModuleArticle.php:71)
  at Contao\ModuleArticle->generate(false)
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:515)
  at Contao\Controller::getArticle(object(ArticleModel), false, false, 'main')
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:377)
  at Contao\Controller::getFrontendModule('0', 'main')
     (vendor/contao/core-bundle/src/Resources/contao/pages/PageRegular.php:168)
  at Contao\PageRegular->prepare(object(PageModel))
     (vendor/contao/core-bundle/src/Resources/contao/pages/PageRegular.php:48)
  at Contao\PageRegular->getResponse(object(PageModel), true)
     (vendor/contao/core-bundle/src/Resources/contao/controllers/FrontendIndex.php:339)
  at Contao\FrontendIndex->renderPage(object(PageModel))
     (vendor/symfony/http-kernel/HttpKernel.php:158)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/http-kernel/HttpKernel.php:80)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/http-kernel/Kernel.php:201)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (web/index.php:31)

**Comment by aschempp** [comment]
I cannot reproduce this problem in the demo (local, not demo.contao.org). Can you please try to reproduce this on the Contao demo?

**Comment by Metis77** [comment]
Had the same issue as described in 4.9.10 and also 4.9.11.
Found out, that I had a custom `mod_login.html5` that just needed to be updated.

<img width="905" alt="Bildschirmfoto 2021-02-03 um 00 14 26" src="https://user-images.githubusercontent.com/1408361/106675805-12df2800-65b6-11eb-8b2b-01b08393fba5.png">

**Comment by paddibr** [comment]
@kartoffelheinz Did you find a solution for the problem?

**Comment by kartoffelheinz** [comment]
@paddibr No, not a solution, just a workaround. We use an early FE Module (earliest time we can be sure there is a login session if we do NOT hit the bug) and do a refresh (via RedirectResponse) if the client did not send a session cookie. But that only works for our specific use case with a closed client group where we can enforce cookie usage.

Since we had the workaround, we moved on to other stuff, so I have not tested Contao 4.13 but will do so shortly in the hopes this finally fixes it.

**Comment by paddibr** [comment]
mhm okay. We have the same error under contao 4.13. With autologin a user need a second refresh. :-/

**Comment by kartoffelheinz** [comment]
Well, thank's for letting me know, spared me some time. But very unfortunate to see this bug has indeed never been fixed, not even in the latest version.

I still do not understand why this ticket has been closed without providing a working solution to this problem.

**Comment by fritzmg** [comment]
> But very unfortunate to see this bug has indeed never been fixed, not even in the latest version.

The bug has been fixed, there has been no new reproduction (see above).
